Add the release note for 1.6.0 (#3436)
line/armeria: site/src/pages/release-notes/1.6.0.mdx

---
date: 2021-04-06
---

## 🌟 New features

> - You can now send and handle the multipart request. #253 #3327

Reviewer: multipart requests and responses? ('the' is not necessary)

- You can now create a <type://StreamMessage> from a `Path` or a `File`. #3344
  ```java
  Path path = Paths.get("...");
  StreamMessage<HttpData> publisher = StreamMessage.of(path);
  ```
- You can now filter or transform values of a <type://StreamMessage> using
  <type://StreamMessage#filter(Predicate)> or <type://StreamMessage#map(Function)>. #3351
  ```java
  // Filter.
  StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
  StreamMessage<Integer> even = source.filter(x -> x % 2 == 0);

  // Transform.
  StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
  StreamMessage<Boolean> isEven = source.map(x -> x % 2 == 0);
  ```
- You can now send a different response depending on the exception. #3209 #3413
  ```java
  ExceptionHandler handler = (ctx, cause) -> {
      if (cause instanceof RequestTimeoutException) {
          // The request timed out!
          return AggregatedHttpResponse.of(...);
      }

      if (cause instanceof IllegalArgumentException) {
          return AggregatedHttpResponse.of(HttpStatus.BAD_REQUEST);
      }

      // Return null to let ExceptionHandler.ofDefault() convert the exception.
      return null;
  }

>   Server.builder().exceptionHandler(handler)...

Reviewer: How about simplifying like this:

```
Server.builder().exceptionHandler((ctx, cause) -> {
    ...
})...
```

  ```
- You can now convert an exception into an <type://RpcResponse> in <type://THttpService>. #3349 #3383
  ```java
  THttpService.builder()
              .addService("hello", helloService)
              .exceptionHandler((ctx, cause) -> {
                  if (cause instanceof IllegalArgumentException) {

>                       return RpcResponse.of(CustomizedException("Bad Request!"));

Reviewer: `new CustomizedException(...)`?

                  }
                  ...
              })
  ```
- You can now set a response timeout and attributes using <type://WebClientRequestPreparation>. #3347 #3357
  ```java
  WebClient client = ...;
  client.prepare()
        .get("/my-service")
        .responseTimeout(Duration.ofSeconds(3))
        .attr(USER_ID, userId)
        .attr(USER_SECRET, secret)
        .execute();
  ```

> - You can now add additional error details to the response when an exception is raised in gRPC service.

Reviewer: - add additional -> specify additional ? (not two 'add's)
- the response -> a response (or a gRPC response)
- a gRPC service

  #3307 #3329
  ```java
  GrpcService.builder()
             .exceptionMapping((cause, metadata) -> {
                 if (throwable instanceof AuthError) {
                     metadata.put(KEY, toMetadata(cause))
                     return Status.UNAUTHENTICATED.withCause(cause);
                 }
                 ...
             })
  ```
- You can now customize the success condition of a metric. #3404 #3410
  ```java
  MetricCollectingService.builder(...)
                         .successFunction((context, log) -> {
                              final int statusCode = log.responseHeaders().status().code();
                              return (statusCode >= 200 && statusCode < 400) || statusCode == 404;
                          });
  ```
- You can now fluently build a <type://DecodingClient> using the <type://DecodingClientBuilder>. #3348 #3372
  ```java
  DecodingClient.builder()
                .autoFillAcceptEncoding(false)
                .strictContentEncoding(true)
                .newDecorator();
  ```
- You can now fluently build an <type://HttpRequest> with a `Publisher<HttpData>`. #3343
  ```java

>   Publisher<HttpData> publisher = StreamMessage.of(...);

Reviewer 2: Just?
```suggestion
  StreamMessage<HttpData> publisher = StreamMessage.of(...);
```

  public Publisher<MyProtobufMessage> protobufJsonSeqPublisher() {
      return StreamMessage.of(MyProtobufMessage.newBuilder()...build(),
                              MyProtobufMessage.newBuilder()...build());
  }

  @Get("/items")
  @ProducesJsonSequences
  public Stream<MyProtobufMessage> protobufJsonSeqPublisher() {
      return Stream.of(MyProtobufMessage.newBuilder()...build(),
                       MyProtobufMessage.newBuilder()...build());
  }
  ```
- You can now customize the default service name of a <type://RequestLog>. #3232 #3366
  ```java
  Server.builder()
        .defaultServiceNaming(ctx -> {
            final ServiceConfig config = ctx.config();
            return config.route().patternString();
       });
  // For a specific service.
  Server.builder()
        .route().path("/")
        .defaultServiceNaming(...)
        ...
  ```
- You can now check if the current request is matched by any routes or not. #3365 #3378
  ```java
  ServerBuilder sb = ...
  sb.decorator((delegate, ctx, req) -> {
      if (ctx.config().route().isFallback()) {
          // This request is not matched any routes.
      }
      return delegate.serve(ctx, req);
  });
  ```

> - The <type://BraveService> does not trace requests to the <type://TransientService> such as

Reviewer: - The BraveService -> BraveService
- to the -> for

  <type://HealthCheckService> anymore. #3382
  - You should specify <type://TransientServiceOption#WITH_TRACING> if you want to trace them.
  ```java
  HealthCheckService.builder()
                    .transientServiceOptions(WITH_TRACING)
                    ...
                    .build();
  ```
- You can now clean up resources by overriding <type://FilteredStreamMessage#onCancellation(Subscriber)>
  when `Subscription.cancel()` is called. #3375
  ```java
  new MyFilteredHttpResponse(res) {
      ...
      @Override
      protected void onCancellation(Subscriber<? super U> subscriber) {
          // Clean up resources.
      }
  }
  ```

> - You can now easily set 'cookie' or 'set-cookie' headers. #3391

Reviewer 2: The getters are also added.

  ```java
  Cookie cookie = Cookie.of("cookie", "value");
  RequestHeaders headers = RequestHeaders.builder(HttpMethod.GET, "/")
                                         .cookie(cookie)
                                         .build();

  Cookie setCookie1 = ...
  Cookie setCookie2 = ...
  ResponseHeaders headers = ResponseHeaders.builder(HttpStatus.OK)
                                           .cookies(setCookie1, setCookie2)
                                           .build();
  ```

> - You can now use `req.root_id` <type://BuiltInProperty> to log the id of <type://RequestContext#root()>.

Reviewer 2: ```suggestion
- You can now use `req.root_id` <type://BuiltInProperty> to log the ID of <type://RequestContext#root()>.
```

  #3429 #3433

> - You can now use Thrift 0.14.0. #3470 #3422

Reviewer 2: ?
```suggestion
- You can now use Thrift 0.14.1. #3470 #3422
```

Reviewer: How about:

    You can now use Thrift 0.14.0 with new `armeria-thrift0.14` module.

> - You can now use OAuth 2.0 related features. #2268 #2840

Reviewer: .. with new `armeria-oauth2` module.

  - [OAuth 2.0 server-side token authorization](https://datatracker.ietf.org/doc/rfc7662/) using
    <type://OAuth2TokenIntrospectionAuthorizer>
  - [Client credential grants](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4) using
    <type://OAuth2ClientCredentialsGrant>
  - [Resource Owner Password Credentials Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-4.3)
    using <type://OAuth2ResourceOwnerPasswordCredentialsGrant>
  - [OAuth 2.0 Token Revocation](https://datatracker.ietf.org/doc/rfc7009/) using <type://TokenRevocation>

  <Tip>

    OAuth 2.0 features are currently experimental, so you should be careful using the feature.

  </Tip>

- You can now use [Jakarta RESTful Web Services](https://jakarta.ee/specifications/restful-ws/)
  on top of Armeria with the `armeria-resteasy` module. #3285 #3296

  <Tip>

    `armeria-resteasy` is currently experimental, so you should be careful using the feature.

  </Tip>

## 📈 Improvements

- Various improvements for documentation. #3237 #3289 #3361 #3390 #3401 #3427

## 🛠️ Bug fixes

- You no longer see `NullPointerException` in <type://HttpResponseDecoder>. #3036 #3407

> - You no longer see 405 when the exact and param path are defined with different HTTP methods. #3330 #3340

Reviewer: 405 -> `405 Method Not Allowed` status

> - You no longer see `Address family not supported by protocol` or `Connection refused` error anymore. #3425

Reviewer: anymore on certain machines with IPv6 enabled.

> - You no longer see escaped Unicode from a JSON response in <type://DocService>. #3396

Reviewer: The Unicode characters like emojis in a JSON response are now rendered correctly in DocService. ?

- You no longer see the wrong response body when the payload violates the protocol or is too large. #3419

> - You can now use the Thrift client and service that is generated by `java:fullcamel` Thrift generator.

Reviewer 2: ```suggestion
- You can now use the Thrift client and service that is generated by [`java:fullcamel`](https://issues.apache.org/jira/browse/THRIFT-2469) option of Thrift compiler.
```

  #3269 #3360 #3369
- `SmartLifecycle` for Armeria server graceful shutdown is only created when the Armeria server is created
  by Spring integration. #3300

> - You can now use [Sealed oneofs](https://scalapb.github.io/docs/sealed-oneofs/) message from ScalaPB
>   using JSON. #3342

Reviewer 2: ```suggestion
- You can now use [Sealed oneofs](https://scalapb.github.io/docs/sealed-oneofs/) message from ScalaPB
  with JSON for gRPC and annotated services. #3342 #3394
```
Add #3394 too?

          ServiceRequestContext ctx, ByteBuf message) {...}
  ```
- <type://GrpcStatusFunction> now extends `BiFunction<Throwable, Metadata, Status>`. #3329
- The `numberSteps` method in <type://WeightRampingUpStrategyBuilder> is now changed to `totalSteps`. #3377

## ⛓ Dependencies

- Bucket4J 4.10.0 → 6.2.0
- Curator 4.3.0 → 5.1.0
  - ZooKeeper 3.5.8 → 3.6.2
- Dropwizard 2.0.18 → 2.0.20
- Dropwizard Metrics 4.1.17 → 4.1.18
- gRPC 1.35.0 → 1.36.1
- Jackson 2.12.0 → 2.12.2
- java-jwt 3.12.1 → 3.14.0
- Jetty 9.4.36 → 9.4.39
- Micrometer 1.6.3 → 1.6.5
- Netty 4.1.58 → 4.1.63
  - Netty TCNative BoringSSL 2.0.36 → 2.0.38
  - Netty io_uring transport 0.0.3 → 0.0.5
- OpenSAML 3.4.5 → 3.4.6
  - Shibboleth java-support 7.5.1 → 7.5.2
- Reactor 3.4.2 → 3.4.4
  - Reactor Kotlin extensions 1.1.2 → 1.1.3
- RxJava 3.0.9 → 3.0.11, 2.2.20 → 2.2.21
- ScalaPb 0.10.10 → 0.11.0
- Spring Boot 2.4.2 → 2.4.4
- Tomcat 9.0.41 → 9.0.44, 8.5.61 → 8.5.64

## 🙇 Thank you

<ThankYou
  usernames={[
    'andrey-tpt',
    'andrew-teirney',
    'anuraaga',
    'eisig',
    'eugene70',
    'ghkim3221',
    'hyangtack',
    'heka1024',
    'heowc',
    'hexoul',
    'ikhoon',
    'JunoJunho',
    'kojilin',

>     'ks-kim',

Reviewer 2: Oops... His GitHub ID was https://github.com/ks-yim . 😱
